dspy/primitives/tool.py

```diff
@@ -147,19 +164,15 @@ def _parse_args(self, **kwargs):
 
     @with_callbacks
     def __call__(self, **kwargs):
-        for k, v in kwargs.items():
-            if k not in self.args:
-                if self.has_kwargs:
-                    # If the tool has kwargs, skip validation for unknown args
-                    continue
-                raise ValueError(f"Arg {k} is not in the tool's args.")
-            try:
-                instance = v.model_dump() if hasattr(v, "model_dump") else v
-                type_str = self.args[k].get("type")
-                if type_str is not None and type_str != "Any":
-                    validate(instance=instance, schema=self.args[k])
-            except ValidationError as e:
-                raise ValueError(f"Arg {k} is invalid: {e.message}")
-
-        parsed_kwargs = self._parse_args(**kwargs)
-        return self.func(**parsed_kwargs)
+        parsed_kwargs = self._validate_and_parse_args(**kwargs)
+        result = self.func(**parsed_kwargs)
+        if asyncio.iscoroutine(result):
+            raise ValueError("You are calling `__call__` on an async tool, please use `acall` instead.")
+        return result
+
+    async def acall(self, **kwargs):
```

Comment:
Q: do we plan to support async tools in ReAct?

Reply:
yes in a followup PR!